### PR TITLE
feat: add support for Redis sintercard command

### LINKED ISSUE #1083: [NEW] Add `SINTERCARD` command aligned Redis
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-kvrocks/issues) and found no similar issues.

### Motivation

add support for [SINTERCARD](https://redis.io/commands/sintercard/) command in Redis, which is available since Redis 7.0.0

### Solution

Since `SINTER` and `SCARD` were implemented already, I think it's quite straightforward to combine those two commands' logic.

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by boatrainlsz**
Add support for Redis sintercard command, this close #1083

### CODE REVIEW

**Comment by tanruixiang** [comment]
Thank you very much for your contribution. Maybe you also need to add the go test, you can refer to `incubator-kvrocks/tests/gocase/unit/type/set/set_test.go`

File: src/commands/redis_cmd.cc [hunk 1]
```diff
@@ -2270,6 +2277,54 @@ class CommandSInter : public Commander {
...
+    Redis::Set set_db(svr->storage_, conn->GetNamespace());
+    int64_t ret = 0;
+    auto s = set_db.InterCard(keys, limit, &ret);
+    if (!s.ok()) {
+      return Status(Status::RedisExecErr, s.ToString());
+    }
+    *output = Redis::Integer(ret);
+    return Status::OK();
+  }
+};
```
Comment (tanruixiang) [inline src/commands/redis_cmd.cc:2492]: You can refer to #1032  to use a more concise parsing approach.

Comment (boatrainlsz) [reply src/commands/redis_cmd.cc:2492]: @tanruixiang Thanks for advice.

File: src/commands/redis_cmd.cc [hunk 2]
```diff
@@ -2270,6 +2277,54 @@ class CommandSInter : public Commander {
   }
 };
 
+class CommandSInterCard : public Commander {
+ public:
+  Status Execute(Server *svr, Connection *conn, std::string *output) override {
```
Comment (PragmaTwice) [inline src/commands/redis_cmd.cc:2448]: I think you can try to split this method to two methods: `Parse` and `Execute`.

File: tests/cppunit/t_set_test.cc [hunk 3]
```diff
@@ -188,6 +188,27 @@ TEST_F(RedisSetTest, Inter) {
   set->Del(k3);
 }
 
+TEST_F(RedisSetTest, InterCard) {
+  int ret;
```
Comment (PragmaTwice) [inline tests/cppunit/t_set_test.cc:192]: Would you like to also write some unit tests in golang? e.g. `tests/gocase/unit/type/...`

File: src/commands/redis_cmd.cc (hunk 1 above)
Comment (PragmaTwice) [reply src/commands/redis_cmd.cc:2492]: It is optional to use the command parser, but I think at least we need to extract the parsing part in this method `Execute` to `Parse`.

File: src/types/redis_set.cc [hunk 4]
```diff
@@ -133,6 +133,17 @@ rocksdb::Status Set::Card(const Slice &user_key, int *ret) {
   return rocksdb::Status::OK();
 }
 
+rocksdb::Status Set::InterCard(const std::vector<Slice> &keys, int64_t limit, int64_t *ret) {
+  std::vector<std::string> members;
+  rocksdb::Status s = Inter(keys, &members);
+  if (!s.ok()) return s;
+  *ret = members.size();
+  if (limit > 0 && *ret > limit) {
```
Comment (torwig) [inline src/types/redis_set.cc:141]: The Redis docs states:

`By default, the command calculates the cardinality of the intersection of all given sets. When provided with the optional LIMIT argument (which defaults to 0 and means unlimited), if the intersection cardinality reaches limit partway through the computation, the algorithm will exit and yield limit as the cardinality. Such implementation ensures a significant speedup for queries where the limit is lower than the actual intersection cardinality.`

I think, your implementation doesn't speed up the process of intersection, it performs an intersection as usual (full) and adjusts the result by comparing it with the `limit` value. In order to truly use the `limit` value, you should use it somewhere inside `Inter()`, passing it as a parameter or introducing a completely separate method.

**Comment by boatrainlsz** [comment]
Everybody, Thank you for all your suggestions, I will recheck related codes.

**Comment by tisonkun** [comment]
Code conflict as we did some refactors. cc @torwig @PragmaTwice could you help with providing the plan on following refactors so that we don't resolve conflicts many times?

**Comment by torwig** [comment]
@tisonkun Resolved. However, this feature is not ready to be merged.

**Comment by PragmaTwice** [comment]
Hi @boatrainlsz , do you still have time to continue this PR?

**Comment by infdahai** [comment]
@PragmaTwice Can I fix this pr because of long-term loss of contact? I don't know how to collaborate with this.

**Comment by PragmaTwice** [comment]
> @PragmaTwice Can I fix this pr because of long-term loss of contact? I don't know how to collaborate with this.

Sure. I think you can pick up this PR.

**Comment by git-hulk** [comment]
Close this PR due to out-of-date and no response.

**Pull request closed without merging**